pkg/collector/scaling_schedule_collector.go

```diff
@@ -0,0 +1,309 @@
+func (c *ScalingScheduleCollector) GetMetrics() ([]CollectedMetric, error) {
+	scalingScheduleInterface, exists, err := c.store.GetByKey(fmt.Sprintf("%s/%s", c.objectReference.Namespace, c.objectReference.Name))
+	if !exists {
+		return nil, ErrScalingScheduleNotFound
+	}
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error retrieving the ScalingSchedule: %s", err.Error())
+	}
+
+	scalingSchedule, ok := scalingScheduleInterface.(*v1.ScalingSchedule)
+	if !ok {
+		return nil, ErrNotScalingScheduleFound
+	}
+	return calculateMetrics(scalingSchedule.Spec.Schedules, c.now(), c.objectReference, c.metric)
+}
+
+// GetMetrics is the main implementation for collector.Collector interface
+func (c *ClusterScalingScheduleCollector) GetMetrics() ([]CollectedMetric, error) {
+	clusterScalingScheduleInterface, exists, err := c.store.GetByKey(c.objectReference.Name)
+	if !exists {
+		return nil, ErrClusterScalingScheduleNotFound
+	}
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error retrieving the ClusterScalingSchedule: %s", err.Error())
+	}
+
+	// The [cache.Store][0] returns the v1.ClusterScalingSchedule items as
+	// a v1.ScalingSchedule when it first lists it. Once the objects are
+	// updated/patched it asserts it correctly to the
+	// v1.ClusterScalingSchedule type. It means we have to handle both
+	// cases.
+	// TODO(jonathanbeber): Identify why it happens and fix in the upstream.
+	//
+	// [0]: https://github.com/kubernetes/client-go/blob/v0.21.1/tools/cache/Store.go#L132-L140
+	var clusterScalingSchedule v1.ClusterScalingSchedule
+	scalingSchedule, ok := clusterScalingScheduleInterface.(*v1.ScalingSchedule)
+	if !ok {
+		css, ok := clusterScalingScheduleInterface.(*v1.ClusterScalingSchedule)
+		if !ok {
+			return nil, ErrNotClusterScalingScheduleFound
+		}
+		clusterScalingSchedule = *css
+	} else {
+		clusterScalingSchedule = v1.ClusterScalingSchedule(*scalingSchedule)
+	}
+
+	return calculateMetrics(clusterScalingSchedule.Spec.Schedules, c.now(), c.objectReference, c.metric)
+}
+
+// Interval returns the interval at which the collector should run.
+func (c *ScalingScheduleCollector) Interval() time.Duration {
+	return c.interval
+}
+
+// Interval returns the interval at which the collector should run.
+func (c *ClusterScalingScheduleCollector) Interval() time.Duration {
+	return c.interval
+}
+
+func calculateMetrics(schedules []v1.Schedule, now time.Time, objectReference custom_metrics.ObjectReference, metric autoscalingv2.MetricIdentifier) ([]CollectedMetric, error) {
+	value := 0
+	for _, schedule := range schedules {
+		switch schedule.Type {
+		case v1.RepeatingSchedule:
+			location, err := time.LoadLocation(schedule.Period.Timezone)
+			if schedule.Period.Timezone == "" || err != nil {
+				location, err = time.LoadLocation(defaultTimeZone)
+				if err != nil {
+					return nil, fmt.Errorf("unexpected error loading default location: %s", err.Error())
+				}
+			}
```

Comment:
If I read this correctly it would silently fallback to `defaultTimezone` in case of invalid timezone (e.g. typo) which could be a surprise for the user.

The `if schedule.Period.Timezone == ""` check does not make much sense as `time.LoadLocation` would return `UTC` for the empty timezone according to the docs https://pkg.go.dev/time#LoadLocation. Besides that I think default timezone (if still used) should be `UTC` as it is common for other timestamps in k8s.

IMO timezone should be either mandatory or empty by default which would result in UTC timezone and the logic should fail fast on error without any fallback unless I am missing something else here.

Reply:
Timezone is required by CRD